extensions/amp-ad/0.1/amp-ad-image.js

```diff
@@ -0,0 +1,94 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {getLifecycleReporter} from '../../../ads/google/a4a/performance';
+import {user} from '../../../src/log';
+import {getBatchManager} from './amp-ad-batch-manager.js';
+import {templatesFor} from '../../../src/template';
+
+/** @const {!string} Tag name for 3P AD implementation. */
+export const TAG_AD_IMAGE = 'amp-ad-image';
+
+export class AmpAdImage extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {string} The base URL of the ad server for this ad */
+    this.url_ = element.getAttribute('data-url');
+
+    /** @private {string} A string identifying this ad slot: the server's responses will be keyed by slot */
+    this.slot_ = element.getAttribute('data-slot');
+    if (!this.slot_.match(/^[0-9a-z]+$/)) {
+      user().error(TAG_AD_IMAGE, 'imagead slot should be alphanumeric: ' +
+            this.slot_);
+    }
+
+    /** @private {Object} Data passed back from the ad server, via the batch manager */
+    this.jsondata_ = null;
+
+    /** @private {AmpAdBatchManager} This will batch up the display of this ad together with others of the same URL */
+    this.batchManager_ = getBatchManager(this, this.url_,
+        'amp-ad[type=imagead]');
+
+    this.lifecycleReporter_ = getLifecycleReporter(this, 'amp');
+    this.lifecycleReporter_.sendPing('adSlotBuilt');
+  }
+
+  /** @override */
+  getPriority() {
+    // Loads ads after other content.
+    return 2;
+  }
+
+  /** @override **/
+  isLayoutSupported(layout) {
+    /** @todo Add proper support for more layouts, and figure out which ones we're permitting */
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * Display an ad in this element. Call this only when we know that the responses have been fetched from the ad server.
+   * The response data can be found in the batch manager's responseData variable, indexed by slot.
+   */
+  showImageAd() {
+    const d = this.batchManager_.responseData[this.slot_];
+    // See if there are any templates in this ad
+    const templates = this.element.querySelectorAll('template');
+    if (templates.length) {
+      // We have a template. Use it to do the rendering.
+      const t = this;
+      templatesFor(this.win).findAndRenderTemplate(this.element, d)
+          .then(function(x) {
+            // Clear out the template and replace it by the rendered version
+            t.element.innerHTML = '';
+            t.element.appendChild(x);
+          });
+    } else {
+      user().error(TAG_AD_IMAGE, 'Missing template in imagead');
+    }
+  }
+  /** @override */
+  layoutCallback() {
+    const t = this;
```

Comment:
The element has to be passed to the batchManager callback somehow. Since you don't like this construct, I've changed it so that the promise from doLayout returns the element as opposed to just 'OK', that can now be passed in the arrow function.